### PR TITLE
abci: clarify connection use in-process

### PR DESCRIPTION
**Pull request by williambanfield**
[rendered](https://github.com/tendermint/spec/blob/wb/abci-connections/spec/abci/abci.md)

For: #316

### CODE REVIEW

File: spec/abci/abci.md [hunk 1]
```diff
@@ -5,27 +5,32 @@ title: Method and Types
...
-The mempool connection is for validating new transactions, before they're
-shared or included in a block.
-
-The info connection is for initialization and for queries from the user.
-
-The snapshot connection is for serving and restoring [state sync snapshots](apps.md#state-sync).
+## Connections
+
+ABCI applications can run either within the same _process_ as the Tendermint 
+state-machine replication engine _or_ as a separate process from the state-machine
```
Comment (creachadair) [inline spec/abci/abci.md:11]: Minor nit: It feels like we're emphasizing different things here.

```suggestion
ABCI applications can run either within the _same_ process as the Tendermint 
state-machine replication engine, or as a _separate_ process from the state-machine
```

File: spec/abci/abci.md [hunk 2]
```diff
@@ -5,27 +5,32 @@ title: Method and Types
...
+replication engine. When run within the same process, Tendermint will call the ABCI
+application methods directly as typical Go method calls. When Tendermint and the
+ABCI application are run as separate processes, Tendermint opens a set of connections
+to the application.
+
+When Tendermint and the application are run as separate processes, Tendermint opens
+four connections for ABCI methods. The connections each handle a subset of the ABCI
+method calls. They are defined as follows:
```
Comment (creachadair) [inline spec/abci/abci.md:19]: There is some duplication between these two paragraphs that I'd suggest collapsing.
```suggestion
application methods directly as Go method calls.

When Tendermint and the ABCI application are run as separate processes, Tendermint
opens four connections to the application for ABCI methods. The connections each 
handle a subset of the ABCI method calls. These subsets are defined as follows:
```

Comment (creachadair) [reply spec/abci/abci.md:19]: Not directly related to this change, but it seems odd to have all of these on separate connections. Is there some rationale for that, which we could summarize? (Made-up hypothetical: "This allows the application to more easily handle different parts of the consensus workflow concurrently" or something like that?)

File: spec/abci/abci.md [hunk 3]
```diff
@@ -5,27 +5,32 @@ title: Method and Types
...
+
+#### **Consensus** connection
+* Handles the `InitChain`, `BeginBlock`, `DeliverTx`, `EndBlock`, and `Commit` method
```
Comment (creachadair) [inline spec/abci/abci.md:22]: I'd suggest putting the "summary" explanation first, and the list of methods second, in each case.

File: spec/abci/abci.md (hunk 1 above)
Comment (williambanfield) [reply spec/abci/abci.md:11]: reread this just now and noticed the same thing.

File: spec/abci/abci.md (hunk 3 above)
Comment (williambanfield) [reply spec/abci/abci.md:22]: done.

File: spec/abci/abci.md (hunk 2 above)
Comment (williambanfield) [reply spec/abci/abci.md:19]: I'm not 100% sure I know the answer. this document is how I learned about how this works. I'd recommend asking in the tendermint slack channel.

**Review by tac0turtle — APPROVED** [review]
looks good, thanks for the additions

**Pull request merged**